core/src/main/java/com/linecorp/armeria/client/endpoint/FileWatcherRegistry.java

```diff
@@ -0,0 +1,234 @@
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client.endpoint;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.nio.file.StandardWatchEventKinds.ENTRY_CREATE;
+import static java.nio.file.StandardWatchEventKinds.ENTRY_DELETE;
+import static java.nio.file.StandardWatchEventKinds.ENTRY_MODIFY;
+import static java.nio.file.StandardWatchEventKinds.OVERFLOW;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.file.FileSystem;
+import java.nio.file.Path;
+import java.nio.file.WatchKey;
+import java.nio.file.WatchService;
+import java.util.Collection;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.MoreObjects;
+
+import com.linecorp.armeria.client.endpoint.FileWatcherRunnable.FileWatchEvent;
+
+/**
+ * Wraps a {@link WatchService} and allows paths to be registered.
+ */
+final class FileWatcherRegistry implements AutoCloseable {
+
+    /**
+     * Context responsible for watching a {@link FileSystem}. Contains a reference to
+     * a {@link WatchService} and a {@link Thread} which continuously watches for changes
+     * in registered file paths.
+     */
+    static class FileSystemWatchContext {
+
+        private final RestartableThread restartableThread;
+        private final WatchService watchService;
+        private final Map<FileWatchRegisterKey, FileWatchEvent> currWatchEventMap =
+                new ConcurrentHashMap<>();
+
+        /**
+         * Creates a new {@link FileSystemWatchContext}.
+         * @param name the name of the thread used to watch for changes
+         * @param watchService the {@link WatchService} used to monitor for changes
+         */
+        FileSystemWatchContext(String name, WatchService watchService) {
+            restartableThread = new RestartableThread(name, () ->
+                    new FileWatcherRunnable(watchService, this));
+            this.watchService = watchService;
+        }
+
+        /**
+         * Starts to watch changes for the path corresponding to the {@link FileWatchRegisterKey}.
+         * When changes are detected, the {@code callback} is invoked. If the {@link WatchService}
+         * isn't running yet, this method starts a thread to start watching the {@link FileSystem}.
+         * @param watchRegisterKey key that contains the path to be watched
+         * @param callback function invoked on file change
+         *
+         * @throws IllegalArgumentException if failed to locate file or failed to start watching
+         */
+        void register(FileWatchRegisterKey watchRegisterKey, Runnable callback) {
+            final Path dirPath = watchRegisterKey.filePath().getParent();
+            checkArgument(dirPath != null, "no parent directory for input path: %s",
```

Comment:
nit. We usualy use `requireNonNull` for null check?

Reply:
I think it has to be `IAE` in this case because the user parameter (`filePath`) is not null but its property (`dirPath`) is.

Reply:
Originally, I thought it would be better to throw an `IllegalArgumentException` over an NPE to let the user know the input is wrong. If this isn't necessary, I'm fine with sticking to convention:) Any thoughts?

Reply:
(It does seem like other input parameters throw NPE though...)
https://github.com/line/armeria/blob/fd6f91f373c03862cbb2a622fd70d8b85c11305c/core/src/main/java/com/linecorp/armeria/client/endpoint/PropertiesEndpointGroup.java#L126
`requireNonNull` might be better as it is consistent with other fields validations

Reply:
oh wait... let me remove the fix for this. I missed @trustin 's comment:(